src/characters/cbd-recipient.ts

```diff
@@ -73,16 +78,36 @@ export class CbdTDecDecrypter {
 
   // Retrieve decryption shares
   public async retrieve(
-    provider: ethers.providers.Web3Provider,
+    web3Provider: ethers.providers.Web3Provider,
     conditionExpr: ConditionExpression,
     variant: number,
     ciphertext: Ciphertext
   ): Promise<DecryptionSharePrecomputed[] | DecryptionShareSimple[]> {
+    const ritualState = await DkgCoordinatorAgent.getRitualState(
+      web3Provider,
+      this.ritualId
+    );
+    if (ritualState !== DkgRitualState.FINALIZED) {
+      throw new Error(
+        `Ritual with id ${this.ritualId} is not finalized. Ritual state is ${ritualState}.`
+      );
+    }
+
+    const isLocallyVerified = await DkgClient.verifyRitual(
```

Comment:
I'm not sure whether this needs to be done every time `retrieve` is called. Perhaps this should be optional i.e. provided as a parameter to `retrieve(...)`. It's unclear to me whether the default value should be `True` of `False` (cc @cygnusv ).

Reply:
I think `true` is a sensible default since the security of the user depends.

Reply:
Maybe. Technically if the ritual made it to completion, then the validation was crowd-sourced to all of the nodes in the cohort and was fine..., no? This is for the unlikely case that the entire cohort was malicious - since any one non-malicious Ursula would have prevented the DKG ritual from completing.

Reply:
We can only know if the participants contributed the correct transcripts once we perform the verification. I don't think there is any other mechanism in place at the moment. For example, how would a non-malicious Ursula prevent the ritual from completing? I don't see this implemented in the `Coordinator` contract.

Notice how without verification it's also possible for malicious Ursulas to shut down any ritual without accountability.

I think in order to address this we need to discuss the scope of configuration sets ("packages") we want to ship and then backtrack how rituals are supposed to work on the client side from there.

Reply:
Would moving the ritual verification step to `CbdTDecDecrypter.create` would be a good temporary solution?

Reply:
> For example, how would a non-malicious Ursula prevent the ritual from completing? I don't see this implemented in the Coordinator contract.

This is done by the honest Ursula providing a correct aggregated transcript that does match the others (from malicious nodes) - https://github.com/nucypher/nucypher-contracts/blob/main/contracts/contracts/coordination/Coordinator.sol#L228. This is why currently the protocol only needs one honest Ursula out of the `n` nodes in the cohort. Subsequently, if all aggregated transcripts match, then ritual state is then set to finalized. Knowing that the state is finalized, means that either all nodes were malicious (probabilistically unlikely for larger-sized cohorts though not 0%) or that everything was good/correct. 

To be clear, local verification is good as an option 👍 , I don't know/am unsure that we want it to be done every time which is the current set-up; should someone be explicit (default parameter is `False` and they have to specifically set it to `True`) about checking it versus being implicit (default to `True` and they have to specifically set it to `False`). In the long term, local verification would not be used at all because the protocol would be even more robust in the future (slashing, proofs, on-chain verification etc.), so having the default be `False` and adopters having to change to `True` could make sense - it's just something we should think about.... what do others think @cygnusv , @KPrasch ?

Side question: how long does the local verification take?

> Would moving the ritual verification step to CbdTDecDecrypter.create would be a good temporary solution?

Interesting. That would seem to be a better spot since the verification would only be done once per encrypter, and is more hidden from the user. That being said, I still have questions about doing local verification in general, but it would definitely be an improvement, and seemingly more hidden from the user. Nice suggestion.

Reply:
> This is done by the honest Ursula providing a correct aggregated transcript 

Thanks for explaining that, I just started working on `Coordinator` changes and didn't have a clear picture there.

> Side question: how long does the local verification take?

The complexity ceiling is a single BLS12-381 pairing, so an order of tens of milliseconds.